### PR TITLE
Visualises requests when in development.

### PR DESCRIPTION
**Pull request by rickharrison**
I took a stab at visualising requests from #33. The one thing that is a little messy is parsing and displaying the json. I ended up keeping a reference to it on the request since you can only call `getRawBody` once per request.

Let me know your thoughts!

screen shot 2016-10-08 at 11 51 34 pm

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -37,6 +55,12 @@ async function run(req, res, fn, onError) {
 
 async function json(req, {limit = '1mb'} = {}) {
   try {
+    if (DEV && req.parsedJson) {
+      return req.parsedJson
+    } else if (DEV && req.jsonError) {
```
Comment (timneutkens) [inline lib/index.js]: Shouldn't we first check if there is an error? 🤔

**Review by timneutkens — APPROVED** [review]
Looks fine 👍, maybe add in a way to pass a custom parse function later on, so you can also log out other stuff. urlencoded forms for example.
For now this will do fine 😄

File: lib/index.js (hunk 1 above)
Comment (rickharrison) [reply lib/index.js]: In theory, there can only be one or the other, but I can switch the order if you like.

**Comment by rauchg** [comment]
This is really beautiful @rickharrison. I think we should have `--log=dev|prod|none` as well that _ignores_ env sniffing.

The prod mode of this would be a _compressed_ version that excludes the body.

**Comment by rauchg** [comment]
(basically something similar to the default express logger)

**Comment by rickharrison** [comment]
Sounds good! I will work on this over my holiday break :)

**Comment by luisrudge** [comment]
This is amazing

**Comment by rickharrison** [comment]
Here is what `--log=prod` looks like:

screen shot 2016-12-26 at 5 40 51 pm

**Review by timneutkens — APPROVED** [review]
Looks good :)

**Comment by timneutkens** [comment]
@rauchg final check?

**Comment by rauchg** [comment]
I'll take a look!

File: bin/micro [hunk 2]
```diff
@@ -26,9 +26,10 @@ let [,, file] = args._;
 
 const help = () => {
   console.log(`Usage: micro [opts] <file>
-  -H  Host to listen on   [0.0.0.0]
-  -p  Port to listen on      [3000]
-  -h  Show this help message`);
+  -H    Host to listen on   [0.0.0.0]
+  -p    Port to listen on      [3000]
+  -h    Show this help message
+  --log dev|prod|none`);
```
Comment (rauchg) [inline bin/micro]: `off` would be better than `none` IMO

Comment (rauchg) [reply bin/micro]: we also need long forms for all of this, in addition to the short form

```
-H, --host
-p, --port
-h, --help
-L, --log
```

File: lib/index.js [hunk 3]
```diff
@@ -5,8 +5,22 @@ const server = require('http').Server
 const getRawBody = require('raw-body')
 const typer = require('media-typer')
 const isStream = require('isstream')
+const chalk = require('chalk')
+const jsome = require('jsome')
 
 const DEV = 'development' === process.env.NODE_ENV
```
Comment (rauchg) [inline lib/index.js:11]: It'd be cool to document in the README that we do the nicer logging upon _explicit_ `NODE_ENV = development` setting

File: lib/index.js [hunk 4]
```diff
@@ -41,6 +63,14 @@ async function run(req, res, fn, onError) {
 
 async function json(req, {limit = '1mb'} = {}) {
   try {
+    if (req.jsonError) {
```
Comment (rauchg) [inline lib/index.js]: I'd like to avoid polluting `req` if at all possible

File: bin/micro (hunk 2 above)
Comment (timneutkens) [reply bin/micro]: I'll add your second comment to my readme PR @rauchg

Comment (timneutkens) [reply bin/micro]: @rauchg PR'ed #122

Comment (timneutkens) [reply bin/micro]: https://github.com/rickharrison/micro/pull/1

File: README.md [hunk 5]
```diff
@@ -250,6 +250,14 @@ function handleErrors (fn) {
 }
 ```
 
+### Logging
+
+Micro provides some built-in logging to help visualise requests. When `NODE_ENV` is set to `development`, micro will display the request along with any json body like this:
```
Comment (timneutkens) [inline README.md]: Remove `some`

File: bin/micro [hunk 6]
```diff
@@ -28,7 +29,8 @@ const help = () => {
   console.log(`Usage: micro [opts] <file>
   -H, --host  Host to listen on   [0.0.0.0]
   -p, --port  Port to listen on      [3000]
-  -h, --help  Show this help message`);
+  -h, --help  Show this help message
+  -L, --log dev|prod|off`);
```
Comment (timneutkens) [inline bin/micro]: Turned this around in my merge commit. First the logging then help, also added `[off]` as default indicator

**Comment by timneutkens** [comment]
@rauchg :shipit: 🚀

**Comment by rauchg** [comment]
@rickharrison @timneutkens I think a few additions would make sense:
- Response time (+ delta). EG: `10:30:00 PM (+30ms)`. Kinda like how `debug` gives you deltas
- Also visualize response objects?
- Not 100% sure, but I believe we can anchor the time to the right hand side with an ANSI escape code? (that'd be neat if possible!)

**Comment by rickharrison** [comment]
Sounds good! I'll take a look at this later tonight.

**Comment by rickharrison** [comment]
@rauchg So I have it logging response objects as well now, but I did it by adding a property to `res` 🙁 

The problem is I'm not sure how to get the response body in the `res.on('finish')` event. Is it possible to see the data chunks from a `http.ServerResponse`? An alternative would be to use the callback parameter in the `res.end()` call. However, I do not know what the current log level is in that method. Do you think it would be better to store the log level in a global variable like `DEV` is from the initial `serve` call? Then instead of attaching to `res`, we could output the response body in the `res.end()`

As another caveat, if `send` or a `return` value is not used, then the logger won't be able to pick up on the response body at all unless I am missing something and getting the data from the `ServerResponse` is possible.

I was unable to find a way to align the time text to the right side. If anyone else has an idea of how to do that with ANSI, please let me know.

Happy to chat more on the zeit slack about this as well!

EDIT: Updated screenshot

screen shot 2017-01-08 at 12 20 35 am

**Comment by rauchg** [comment]
that looks really amazing! One way we could capture the response I assume is by patching the internal write method of the response object? The one that both `res.end` and `res.write` call

**Comment by rickharrison** [comment]
Yea, we definitely could do that. I can take a stab and patching those methods inside of the `logRequest` method.

**Comment by rauchg** [comment]
Also, I have a slightly different take on this now:

```jsx
{
  "dependencies": {
     "micro": "latest"
  },
  "devDependencies": {
     "micro-dev": "latest"
  }
}
```

- `micro-dev` would have `micro` in `peerDependencies`, rather than include it. 
- It works with the exact same signature as `micro`, but…
- We can add more options related to logging, for example whether to include headers, if you want to log only requests or only responses, etc.
- We get rid of `NODE_ENV` sniffing. If you run `micro-dev`, it logs. 
- Works really nicely for `package.json`. `start` can run `micro`, `dev` can run `micro-dev`

This way, `micro` remains lean both in download time, LOC/maintainability & performance!

**Comment by rauchg** [comment]
https://github.com/zeit/micro-dev

**Comment by onbjerg** [comment]
@rauchg The repository is empty -- forgot to push or do you need a PR? 😊

**Comment by timneutkens** [comment]
I'm working on it @onbjerg 😄

**Comment by onbjerg** [comment]
@timneutkens I couldn't wait so I implemented it as a simple wrapper function instead of a CLI tool: https://github.com/onbjerg/micro-visualize

**Comment by timneutkens** [comment]
@onbjerg micro-dev is most likely coming today :)

**Comment by onbjerg** [comment]
@timneutkens 😨

**Comment by iamstarkov** [comment]
@timneutkens https://github.com/zeit/micro-dev is 404

**Comment by onbjerg** [comment]
https://github.com/onbjerg/micro-visualize is still an option

**Pull request closed without merging**